homeassistant/components/ondilo_ico/sensor.py

```diff
@@ -0,0 +1,186 @@
+    DEVICE_CLASS_SIGNAL_STRENGTH,
+    DEVICE_CLASS_TEMPERATURE,
+    PERCENTAGE,
+    TEMP_CELSIUS,
+)
+from homeassistant.helpers.entity import Entity
+from homeassistant.helpers.update_coordinator import DataUpdateCoordinator, UpdateFailed
+
+from .const import DOMAIN
+
+SENSOR_TYPES = {
+    "temperature": [
+        "Temperature",
+        TEMP_CELSIUS,
+        "mdi:thermometer",
+        DEVICE_CLASS_TEMPERATURE,
+    ],
+    "orp": ["Oxydo Reduction Potential", "mV", "mdi:pool", None],
+    "ph": ["pH", "", "mdi:pool", None],
+    "tds": ["TDS", CONCENTRATION_PARTS_PER_MILLION, "mdi:pool", None],
+    "battery": ["Battery", PERCENTAGE, "mdi:battery", DEVICE_CLASS_BATTERY],
+    "rssi": [
+        "RSSI",
+        PERCENTAGE,
+        "mdi:wifi-strength-2",
+        DEVICE_CLASS_SIGNAL_STRENGTH,
+    ],
+    "salt": ["Salt", "mg/L", "mdi:pool", None],
+}
+
+SCAN_INTERVAL = timedelta(hours=1)
+
+_LOGGER = logging.getLogger(__name__)
+
+
+async def async_setup_entry(hass, entry, async_add_entities):
+    """Set up the Ondilo ICO sensors."""
+
+    api = hass.data[DOMAIN][entry.entry_id]
+
+    async def async_update_data():
+        """Fetch data from API endpoint.
+
+        This is the place to pre-process the data to lookup tables
+        so entities can quickly look up their data.
+        """
+        try:
+            pools = await hass.async_add_executor_job(api.get_pools)
+            for pool in pools:
+                pool["ICO"] = await hass.async_add_executor_job(
+                    api.get_ICO_details, pool["id"]
+                )
+                pool["devices"] = await hass.async_add_executor_job(
+                    api.get_last_pool_measures, pool["id"]
+                )
+
+            return pools
+
+        except OndiloError as err:
+            raise UpdateFailed(f"Error communicating with API: {err}") from err
+
+    coordinator = DataUpdateCoordinator(
+        hass,
+        _LOGGER,
+        # Name of the data. For logging purposes.
+        name="sensor",
+        update_method=async_update_data,
+        # Polling interval. Will only be polled if there are subscribers.
+        update_interval=SCAN_INTERVAL,
+    )
+
+    # Fetch initial data so we have data when entities subscribe
+    await coordinator.async_refresh()
+
+    entities = []
+    for poolidx, pool in enumerate(coordinator.data):
+        for devidx, _ in enumerate(pool["devices"]):
+            entities.append(OndiloICO(coordinator, poolidx, devidx))
+
+    async_add_entities(entities)
+
+
+class OndiloICO(Entity):
+    """Representation of a Sensor."""
+
+    def __init__(self, coordinator: DataUpdateCoordinator, poolidx: int, devidx: int):
+        """Initialize sensor entity with data from coordinator."""
+        self._coordinator = coordinator
+        self._devidx = devidx
+        self._poolidx = poolidx
+        self._unique_id = (
+            f"{self._pooldata()['ICO']['serial_number']}-{self._devdata()['data_type']}"
```

Comment:
Did you consider using `uuid` here? Curious how you think about that compared to serial_number.

Reply:
Well, this couple uniquely identifies a sensor, so I thought I could use that and that it could be easier to troubleshoot. Not so sure about that any more but it doesn't harm I think. Would you rather use a UUID? If that's usually what's recommended, that's not hard to change (although might be an issue for compatibility with those already using this component as a custom_component)

Reply:
You'd know best, I was just making observations reading the API docs.